### ISSUE TITLE
Notation defined with '@' does not accept arguments

### ISSUE THREAD

**Issue by cpitclaudel** [coq/coq#10546]
##### Description of the problem

Definition a notation for a partially applied term only works if the arguments of the term were not implicit.  This fails:

```coq
Definition pl {x y} := x + y.
Notation p := (@pl 0).
Fail Check (p 3).
(* The expression "p" of type "nat" cannot be applied to the term "3" : "nat" *)
```

… but this works:

```coq
Definition pl' x y := x + y.
Notation p' := (@pl' 0).
Check (p' 3).
```

is this expected?

#### Coq Version

Coq version 8.10+alpha (May 2019), running on GNU/Linux Mint 19.1.

**Comment by JasonGross** [comment]
This is a feature, not a bug (but if it's not documented, that's a bug).  The feature, I believe, is that if the notation is bound to a constant followed by some prefix of its argument list, then the notation picks up the implicit status of the constant on the remaining arguments.   (I believe you can write `@p 3`, for example.)

**Comment by cpitclaudel** [comment]
Neat! I didn't realize that.
Is there a way to do the same with definitions? That's something I've often wanted when defining aliases.

**Comment by JasonGross** [comment]
Nope, no way to do it with definitions.  At least as far as I know

**Comment by herbelin** [comment]
@cpitclaudel: can you give more precisions about your use case?

Conversely, the idea has already been proposed to support assigning implicit arguments to abbreviations (such as your `Notation p := (@pl 0)`), and more generally that `Arguments` work for abbreviations, and maybe even for notations (of the `Notation "..."` form) whose interpretation is a partially applied constant. Would it be of interest for you (or @JasonGross)?.

**Comment by cpitclaudel** [comment]
@herbelin the use case would be defining partially aliases without having to repeat the arguments, with typechecking.

Let's say I have a compiler with 3 type parameters: `var_t`, `annot_t`, `name_t`.  I also have compilation functions `desugar`, `typecheck`, `compile`.

Repeatedly through my codebase I have sections where I re-add these three parameters as `Context` variables, followed by notations to specialize my three compilation functions.

Sometimes I add an extra argument, and I need to update all these sections and all these notations.   Often I forget one, and I get an error — but not when the notation is defined: instead, I get the error when the notation is used.

Of course I can use a definition instead, but the problem is that I have to re-mention all the arguments of each function that I'm partly specializing, to preserve implicits.

I can give a more concrete example if needed.

**Comment by herbelin** [comment]
@cpitclaudel : Thanks for the details. Then it seems important for your case that notations preserve implicit arguments. But your initial message seems to say the converse, that you were expecting the notation not to preserve the implicit arguments. Or maybe, it is that you expect @ in the rhs of the notation to stop inheriting the implicit arguments of the underlying constant? This could be implementable, see #11091, up to the worry that some partial applications cannot be expressed without using @ - in the presence of maximal implicit arguments, so we would need another way to tell that we want to inherit implicit arguments.

Told otherwise, what is non satisfactory in using, in the section, `Notation compile := (compile var_t:=var_t) (annot_t:=annot_t) (name_t:=ame_t))` or `Notation compile := (@compile var_t annot_t  name_t)`?

**Comment by cpitclaudel** [comment]
Sorry, I thought you were asking for the use case of having a Definition inheriting implicits.

> Told otherwise, what is non satisfactory in using, in the section, …

The second one breaks every time the arguments of the function change, and it breaks at the wrong place (when the notation is used, not when it's defined).  The first one is tricky to use when there's a mix of implicit and explicit arguments.

**Comment by cpitclaudel** [comment]
Here's a concrete example of what most of my code looks like: https://github.com/coq/coq/issues/11110

**Comment by herbelin** [comment]
> Sorry, I thought you were asking for the use case of having a Definition inheriting implicits.

This was not what I was precisely asking but now I understand what you meant by "Is there a way to do the same with definitions?". So, thanks.

> The second one breaks every time the arguments of the function change, and it breaks at the wrong place (when the notation is used, not when it's defined).

Maybe `Notation` could be type-checked even if the inferred holes are then discarded (the same way e.g. the pattern is typed in tactic `change pat with term`). That may be tricky to do but worth a try.

But, somehow, I'm still confused about whether you always intend an incantation of the form `Notation compile := (@compile var_t annot_t  name_t)` to inherit the implicit arguments of `compile` or not. Conversely, are there cases in your practice where you want a notation of the form `Notation foo := @bar ...` or `Notation foo args := @bar ...` to not inherit the implicit arguments???

**Comment by cpitclaudel** [comment]
I think the inheritance is good :)  I just hadn't understood how it worked.
Now what I want is a version of definition that works in the same way ^^, or alternatively a version of "notation" that does some checks like a definition.

**Comment by herbelin** [comment]
@cpitclaudel: OK
